spring-boot-project/spring-boot-autoconfigure/src/test/java/org/springframework/boot/autoconfigure/pulsar/PulsarAutoConfigurationTests.java

```diff
@@ -179,6 +216,49 @@ void injectsExpectedBeans() {
 					.hasFieldOrPropertyWithValue("topicResolver", context.getBean(TopicResolver.class)));
 		}
 
+		@ParameterizedTest
+		@ValueSource(booleans = { true, false })
+		@SuppressWarnings("unchecked")
+		void whenHasUserDefinedCustomizersAppliesInCorrectOrder(boolean cachingEnabled) {
+			this.contextRunner
+				.withPropertyValues("spring.pulsar.producer.cache.enabled=" + cachingEnabled,
+						"spring.pulsar.producer.name=fromPropsCustomizer")
+				.withUserConfiguration(ProducerBuilderCustomizersConfig.class)
+				.run((context) -> {
+					DefaultPulsarProducerFactory<?> producerFactory = context
+						.getBean(DefaultPulsarProducerFactory.class);
+					List<ProducerBuilderCustomizer<Object>> customizersOnFactory = (List<ProducerBuilderCustomizer<Object>>) ReflectionTestUtils
+						.getField(producerFactory, "defaultConfigCustomizers");
+					ProducerBuilder<Object> builder = mock(ProducerBuilder.class);
+					customizersOnFactory.forEach((customizer) -> customizer.customize(builder));
+					// Order of customizers should be default props customizer
```

Comment:
I added this "customizer execution order" test for each component. The only way I found to do so was via the customizers all executing the same method w/ different values and verify the order. 

I am looking at creating a `CompositeCustomizer` that takes in a default + list of user defined customizers. That way we could test that functionality once and then just make sure that the composite is invoked.

Reply:
I've added a test utility to extract the common testing logic. I don't think we need a `CompositeCustomizer` in order to merge.